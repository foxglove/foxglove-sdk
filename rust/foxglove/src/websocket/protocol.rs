pub mod client;
pub mod server;

/// Returns true if the given `schema_encoding` is one of the types that is known to require binary
/// schema data (i.e. `protobuf` and `flatbuffer`). These require base64-encoding/decoding to be
/// sent via JSON messages on the websocket connection.
fn is_known_binary_schema_encoding(schema_encoding: &str) -> bool {
    schema_encoding == "protobuf" || schema_encoding == "flatbuffer"
}
