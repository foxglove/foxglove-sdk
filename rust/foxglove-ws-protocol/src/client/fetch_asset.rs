use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Fetch asset message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#fetch-asset>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "fetchAsset", rename_all = "camelCase")]
pub struct FetchAsset {
    /// Asset URI.
    pub uri: String,
    /// Request ID.
    pub request_id: u32,
}

impl JsonMessage for FetchAsset {}

#[cfg(test)]
mod tests {
    use crate::client::ClientMessage;

    use super::*;

    fn message() -> FetchAsset {
        FetchAsset {
            uri: "package://foxglove/example.urdf".to_string(),
            request_id: 42,
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ClientMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ClientMessage::FetchAsset(orig));
    }
}
