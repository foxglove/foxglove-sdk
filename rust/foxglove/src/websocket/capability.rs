use super::ws_protocol::server::server_info;

/// A capability that a websocket server can support.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Capability {
    /// Allow clients to advertise channels to send data messages to the server.
    ClientPublish,
    /// Allow clients to get & set parameters, and subscribe to updates.
    Parameters,
    /// Inform clients about the latest server time.
    ///
    /// This allows accelerated, slowed, or stepped control over the progress of time. If the
    /// server publishes time data, then timestamps of published messages must originate from the
    /// same time source.
    Time,
    /// Allow clients to call services.
    Services,
    /// Allow clients to request assets. If you supply an asset handler to the server, this
    /// capability will be advertised automatically.
    Assets,
    /// Allow clients to subscribe and make connection graph updates
    ConnectionGraph,
    /// Indicates that the server is sending data within a fixed time range. This requires the
    /// server to specify the `data_start_time` and `data_end_time` fields in its `ServerInfo` message.
    RangedPlayback,
}

impl Capability {
    pub(crate) fn as_protocol_capabilities(&self) -> &'static [server_info::Capability] {
        match self {
            Self::ClientPublish => &[server_info::Capability::ClientPublish],
            Self::Parameters => &[
                server_info::Capability::Parameters,
                server_info::Capability::ParametersSubscribe,
            ],
            Self::Time => &[server_info::Capability::Time],
            Self::Services => &[server_info::Capability::Services],
            Self::Assets => &[server_info::Capability::Assets],
            Self::ConnectionGraph => &[server_info::Capability::ConnectionGraph],
            Self::RangedPlayback => &[server_info::Capability::RangedPlayback],
        }
    }
}
