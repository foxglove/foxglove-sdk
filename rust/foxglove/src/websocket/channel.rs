use super::ws_protocol::{client, schema, server};
use crate::{RawChannel, Schema};

/// A client channel ID.
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct ClientChannelId(u32);

impl ClientChannelId {
    /// Creates a new client channel ID.
    pub fn new(id: u32) -> Self {
        Self(id)
    }
}

impl From<ClientChannelId> for u32 {
    fn from(id: ClientChannelId) -> u32 {
        id.0
    }
}

impl From<ClientChannelId> for u64 {
    fn from(id: ClientChannelId) -> u64 {
        id.0.into()
    }
}

impl std::fmt::Display for ClientChannelId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// Information about a channel advertised by the client
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ClientChannel {
    /// An identifier for this channel, assigned by the client
    pub id: ClientChannelId,
    /// The topic name for this channel
    pub topic: String,
    /// The encoding of messages on this channel
    pub encoding: String,
    /// The name of the schema for this channel
    pub schema_name: String,
    /// The encoding of the schema data
    pub schema_encoding: Option<String>,
    /// May or may not be a UTF-8 string depending on the schema_encoding.
    pub schema: Option<Vec<u8>>,
}

impl TryFrom<client::advertise::Channel<'_>> for ClientChannel {
    type Error = schema::DecodeError;

    fn try_from(ch: client::advertise::Channel) -> Result<Self, Self::Error> {
        let schema = match ch.decode_schema() {
            Ok(schema) => Some(schema),
            Err(schema::DecodeError::MissingSchema) => None,
            Err(e) => return Err(e),
        };
        Ok(Self {
            id: ClientChannelId::new(ch.id),
            topic: ch.topic.to_string(),
            encoding: ch.encoding.to_string(),
            schema_name: ch.schema_name.to_string(),
            schema_encoding: ch.schema_encoding.map(|s| s.to_string()),
            schema,
        })
    }
}

impl<'a> TryFrom<&'a RawChannel> for server::advertise::Channel<'a> {
    type Error = schema::EncodeError;

    fn try_from(ch: &'a RawChannel) -> Result<Self, Self::Error> {
        let mut builder = Self::builder(ch.id().into(), ch.topic(), ch.message_encoding());
        if let Some(schema) = ch.schema() {
            builder = builder.with_schema(schema.into());
        }
        builder.build()
    }
}

impl<'a> From<&'a Schema> for schema::Schema<'a> {
    fn from(schema: &'a Schema) -> Self {
        Self::new(&schema.name, &schema.encoding, schema.data.clone())
    }
}
