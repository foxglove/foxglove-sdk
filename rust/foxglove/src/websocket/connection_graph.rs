use crate::websocket::protocol::server::{
    ConnectionGraphDiff, NewAdvertisedService, NewPublishedTopic, NewSubscribedTopic,
};
use std::collections::{HashMap, HashSet};

use super::ClientId;

/// A HashMap where the keys are the topic or service name and the value is a set of string ids.
type MapOfSets = HashMap<String, HashSet<String>>;

/// The connection graph data. Requires capability [`ConnectionGraph`](super::Capability::ConnectionGraph).
/// See <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#connection-graph-update>
#[derive(Debug, Default, Clone)]
pub struct ConnectionGraph {
    /// A map of active topic names to the set of string publisher ids.
    published_topics: MapOfSets,
    /// A map of active topic names to the set of string subscriber ids.
    subscribed_topics: MapOfSets,
    /// A map of active service names to the set of string provider ids.
    advertised_services: MapOfSets,
    /// A set of subscribers.
    subscribers: HashSet<ClientId>,
}

impl ConnectionGraph {
    /// Create a new, empty connection graph.
    pub fn new() -> Self {
        Self::default()
    }

    /// Set a published topic and its associated publisher ids.
    /// Overwrites any existing topic with the same name.
    pub fn set_published_topic(
        &mut self,
        topic: impl Into<String>,
        publisher_ids: impl IntoIterator<Item = impl Into<String>>,
    ) {
        self.published_topics.insert(
            topic.into(),
            HashSet::from_iter(publisher_ids.into_iter().map(|id| id.into())),
        );
    }

    /// Set a subscribed topic and its associated subscriber ids.
    /// Overwrites any existing topic with the same name.
    pub fn set_subscribed_topic(
        &mut self,
        topic: impl Into<String>,
        subscriber_ids: impl IntoIterator<Item = impl Into<String>>,
    ) {
        self.subscribed_topics.insert(
            topic.into(),
            HashSet::from_iter(subscriber_ids.into_iter().map(|id| id.into())),
        );
    }

    /// Set an advertised service and its associated provider ids.
    /// Overwrites any existing service with the same name.
    pub fn set_advertised_service(
        &mut self,
        service: impl Into<String>,
        provider_ids: impl IntoIterator<Item = impl Into<String>>,
    ) {
        self.advertised_services.insert(
            service.into(),
            HashSet::from_iter(provider_ids.into_iter().map(|id| id.into())),
        );
    }

    /// Adds a connection graph subscription for the client.
    ///
    /// Returns false if this client is already subscribed.
    pub(crate) fn add_subscriber(&mut self, client_id: ClientId) -> bool {
        self.subscribers.insert(client_id)
    }

    /// Removes a connection graph subscription for the client.
    ///
    /// Returns false if this client is already unsubscribed.
    pub(crate) fn remove_subscriber(&mut self, client_id: ClientId) -> bool {
        self.subscribers.remove(&client_id)
    }

    /// Returns true if the graph has subscribers.
    pub(crate) fn has_subscribers(&self) -> bool {
        !self.subscribers.is_empty()
    }

    /// Returns true if the client is a subscriber.
    pub(crate) fn is_subscriber(&self, client_id: ClientId) -> bool {
        self.subscribers.contains(&client_id)
    }

    /// Computes a diff of this connection graph against the other.
    pub(crate) fn diff<'a>(&self, other: &'a ConnectionGraph) -> ConnectionGraphDiff<'a> {
        let mut diff = ConnectionGraphDiff::new();

        // Get new or changed published topics
        for (name, publisher_ids) in &other.published_topics {
            if let Some(self_publisher_ids) = self.published_topics.get(name) {
                if self_publisher_ids == publisher_ids {
                    // No change
                    continue;
                }
            }

            diff.published_topics.push(NewPublishedTopic {
                name,
                publisher_ids,
            });
        }

        // Get new or changed subscribed topics
        for (name, subscriber_ids) in &other.subscribed_topics {
            if let Some(self_subscriber_ids) = self.subscribed_topics.get(name) {
                if self_subscriber_ids == subscriber_ids {
                    // No change
                    continue;
                }
            }

            diff.subscribed_topics.push(NewSubscribedTopic {
                name,
                subscriber_ids,
            });
        }

        // Get new or changed advertised services
        for (name, provider_ids) in &other.advertised_services {
            if let Some(self_provider_ids) = self.advertised_services.get(name) {
                if self_provider_ids == provider_ids {
                    // No change
                    continue;
                }
            }

            diff.advertised_services
                .push(NewAdvertisedService { name, provider_ids });
        }

        // Get removed advertised services
        for name in self.advertised_services.keys() {
            if !other.advertised_services.contains_key(name) {
                diff.removed_services.push(name.clone());
            }
        }

        // Get the topics from both published_topics and subscribed_topics that are no longer in either
        for name in self
            .published_topics
            .keys()
            .chain(self.subscribed_topics.keys())
        {
            if other.published_topics.contains_key(name) {
                continue;
            }
            if other.subscribed_topics.contains_key(name) {
                continue;
            }
            diff.removed_topics.insert(name.clone());
        }

        diff
    }

    /// Returns a `ConnectionGraphUpdate` message for the initial state of the graph.
    ///
    /// See: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#connection-graph-update>
    pub(crate) fn as_initial_update(&self) -> String {
        ConnectionGraph::default().diff(self).to_json()
    }

    /// Replaces the connection graph content.
    ///
    /// The set of subscribers is not modified.
    ///
    /// Returns a `ConnectionGraphUpdate` message as a JSON string.
    /// See: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#connection-graph-update>
    pub(crate) fn update(&mut self, new: ConnectionGraph) -> String {
        let diff = self.diff(&new).to_json();
        self.published_topics = new.published_topics;
        self.subscribed_topics = new.subscribed_topics;
        self.advertised_services = new.advertised_services;
        diff
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::{json, Value};

    fn assert_json_eq(left: String, right: Value) {
        let left: Value = serde_json::from_str(&left).unwrap();
        assert_eq!(left, right);
    }

    #[test]
    fn test_empty_update() {
        let mut graph = ConnectionGraph::new();
        let updated = ConnectionGraph::new();
        let diff = graph.update(updated);

        assert_json_eq(
            diff,
            json!({
                "op": "connectionGraphUpdate",
                "publishedTopics": [],
                "subscribedTopics": [],
                "advertisedServices": [],
                "removedTopics": [],
                "removedServices": []
            }),
        );
    }

    #[test]
    fn test_new_published_topic() {
        let mut graph = ConnectionGraph::new();
        let mut updated = ConnectionGraph::new();

        updated.published_topics.insert(
            "topic1".to_string(),
            HashSet::from(["publisher1".to_string()]),
        );

        let diff = graph.update(updated);

        assert_json_eq(
            diff,
            json!({
                "op": "connectionGraphUpdate",
                "publishedTopics": [{
                    "name": "topic1",
                    "publisherIds": ["publisher1"]
                }],
                "subscribedTopics": [],
                "advertisedServices": [],
                "removedTopics": [],
                "removedServices": []
            }),
        );
    }

    #[test]
    fn test_removed_topic() {
        let mut graph = ConnectionGraph::new();
        graph.published_topics.insert(
            "topic1".to_string(),
            HashSet::from(["publisher1".to_string()]),
        );

        let updated = ConnectionGraph::new();
        let diff = graph.update(updated);

        assert_json_eq(
            diff,
            json!({
                "op": "connectionGraphUpdate",
                "publishedTopics": [],
                "subscribedTopics": [],
                "advertisedServices": [],
                "removedTopics": ["topic1"],
                "removedServices": []
            }),
        );
    }

    #[test]
    fn test_changed_publishers() {
        let mut graph = ConnectionGraph::new();
        graph.published_topics.insert(
            "topic1".to_string(),
            HashSet::from(["publisher1".to_string()]),
        );

        let mut updated = ConnectionGraph::new();
        updated.published_topics.insert(
            "topic1".to_string(),
            HashSet::from(["publisher2".to_string()]),
        );

        let diff = graph.update(updated);

        assert_json_eq(
            diff,
            json!({
                "op": "connectionGraphUpdate",
                "publishedTopics": [{
                    "name": "topic1",
                    "publisherIds": ["publisher2"]
                }],
                "subscribedTopics": [],
                "advertisedServices": [],
                "removedTopics": [],
                "removedServices": []
            }),
        );
    }

    #[test]
    fn test_service_changes() {
        let mut graph = ConnectionGraph::new();
        graph.advertised_services.insert(
            "service1".to_string(),
            HashSet::from(["provider1".to_string()]),
        );

        let mut updated = ConnectionGraph::new();
        updated.advertised_services.insert(
            "service2".to_string(),
            HashSet::from(["provider2".to_string()]),
        );

        let diff = graph.update(updated);

        assert_json_eq(
            diff,
            json!({
                "op": "connectionGraphUpdate",
                "publishedTopics": [],
                "subscribedTopics": [],
                "advertisedServices": [{
                    "name": "service2",
                    "providerIds": ["provider2"]
                }],
                "removedTopics": [],
                "removedServices": ["service1"]
            }),
        );
    }

    #[test]
    fn test_complex_update() {
        let mut graph = ConnectionGraph::new();
        graph.published_topics.insert(
            "topic1".to_string(),
            HashSet::from(["publisher1".to_string()]),
        );
        graph.subscribed_topics.insert(
            "topic1".to_string(),
            HashSet::from(["subscriber1".to_string()]),
        );
        graph.advertised_services.insert(
            "service1".to_string(),
            HashSet::from(["provider1".to_string()]),
        );

        let mut updated = ConnectionGraph::new();
        updated.published_topics.insert(
            "topic2".to_string(),
            HashSet::from(["publisher2".to_string()]),
        );
        updated.subscribed_topics.insert(
            "topic2".to_string(),
            HashSet::from(["subscriber2".to_string()]),
        );
        updated.advertised_services.insert(
            "service2".to_string(),
            HashSet::from(["provider2".to_string()]),
        );

        let diff = graph.update(updated);

        assert_json_eq(
            diff,
            json!({
                "op": "connectionGraphUpdate",
                "publishedTopics": [{
                    "name": "topic2",
                    "publisherIds": ["publisher2"]
                }],
                "subscribedTopics": [{
                    "name": "topic2",
                    "subscriberIds": ["subscriber2"]
                }],
                "advertisedServices": [{
                    "name": "service2",
                    "providerIds": ["provider2"]
                }],
                "removedTopics": ["topic1"],
                "removedServices": ["service1"]
            }),
        );
    }
}
