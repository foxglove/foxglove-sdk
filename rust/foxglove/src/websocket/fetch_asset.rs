use std::{future::Future, sync::Arc};

use bytes::Bytes;

use super::{Client, SemaphoreGuard};

/// The result of a fetch asset request.
pub type FetchAssetResult = Result<Bytes, String>;

/// A handler to respond to fetch asset requests.
///
/// See: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#fetch-asset>
pub trait AssetHandler: Send + Sync + 'static {
    /// Fetch an asset with the given uri and return it via the responder.
    /// Fetch should not block, it should call `runtime.spawn`
    /// or `runtime.spawn_blocking` to do the actual work.
    fn fetch(&self, _uri: String, _responder: AssetResponder);
}

pub(crate) struct BlockingAssetHandlerFn<F>(pub Arc<F>);

impl<F> AssetHandler for BlockingAssetHandlerFn<F>
where
    F: Fn(Client, String) -> FetchAssetResult + Send + Sync + 'static,
{
    fn fetch(&self, uri: String, responder: AssetResponder) {
        let func = self.0.clone();
        tokio::task::spawn_blocking(move || {
            let result = (func)(responder.client(), uri);
            responder.respond(result);
        });
    }
}

pub(crate) struct AsyncAssetHandlerFn<F>(pub Arc<F>);

impl<F, Fut> AssetHandler for AsyncAssetHandlerFn<F>
where
    F: Fn(Client, String) -> Fut + Send + Sync + 'static,
    Fut: Future<Output = FetchAssetResult> + Send + 'static,
{
    fn fetch(&self, uri: String, responder: AssetResponder) {
        let func = self.0.clone();
        tokio::spawn(async move {
            let result = (func)(responder.client(), uri).await;
            responder.respond(result);
        });
    }
}

/// Wraps a weak reference to a Client and provides a method
/// to respond to the fetch asset request from that client.
pub struct AssetResponder {
    /// The client requesting the asset.
    client: Client,
    request_id: u32,
    _guard: SemaphoreGuard,
}

impl AssetResponder {
    /// Create a new asset responder for a fetch asset request.
    pub(crate) fn new(client: Client, request_id: u32, guard: SemaphoreGuard) -> Self {
        Self {
            client,
            request_id,
            _guard: guard,
        }
    }

    /// Return a clone of the Client.
    pub fn client(&self) -> Client {
        self.client.clone()
    }

    /// Send an response to the client.
    pub fn respond(self, result: FetchAssetResult) {
        self.client.send_asset_response(result, self.request_id);
    }
}
