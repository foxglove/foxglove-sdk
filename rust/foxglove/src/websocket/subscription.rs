use super::ws_protocol::client::subscribe;
use crate::ChannelId;

/// A client subscription with typed IDs.
// REVIEW: Does this struct need to exist?
pub(crate) struct Subscription {
    pub channel_id: ChannelId,
}
impl From<subscribe::Subscription> for Subscription {
    fn from(value: subscribe::Subscription) -> Self {
        Self {
            channel_id: ChannelId::new(value.channel_id),
        }
    }
}
