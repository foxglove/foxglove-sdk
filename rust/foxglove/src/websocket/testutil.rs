use std::time::Duration;

use futures_util::{SinkExt, StreamExt};
use tokio::net::TcpStream;
use tokio_native_tls::native_tls;
use tokio_tungstenite::tungstenite::client::IntoClientRequest;
use tokio_tungstenite::tungstenite::http::HeaderValue;
use tokio_tungstenite::tungstenite::{self, Message};
use tokio_tungstenite::{MaybeTlsStream, WebSocketStream};

use super::handshake::SUBPROTOCOL;
use super::ws_protocol::server::ServerMessage;
use super::ws_protocol::ParseError;

#[derive(Debug, thiserror::Error)]
pub enum RecvError {
    #[error("unexpected end of stream")]
    UnexpectedEndOfStream,
    #[error(transparent)]
    ParseError(#[from] ParseError),
    #[error(transparent)]
    Tungstenite(#[from] tungstenite::Error),
    #[error(transparent)]
    Timeout(#[from] tokio::time::error::Elapsed),
}

pub struct WebSocketClient {
    stream: WebSocketStream<MaybeTlsStream<TcpStream>>,
}

impl WebSocketClient {
    /// Connects to a server and validates the handshake response.
    pub async fn connect(addr: impl std::fmt::Display) -> Self {
        Self::do_connect(addr, false).await
    }

    pub async fn connect_secure(addr: impl std::fmt::Display) -> Self {
        Self::do_connect(addr, true).await
    }

    pub async fn do_connect(addr: impl std::fmt::Display, use_tls: bool) -> Self {
        let protocol = if use_tls { "wss" } else { "ws" };
        let mut request = format!("{protocol}://{addr}/")
            .into_client_request()
            .expect("Failed to build request");

        request.headers_mut().insert(
            "sec-websocket-protocol",
            HeaderValue::from_static(SUBPROTOCOL),
        );

        let (stream, response) = if use_tls {
            // For tests, ignore TLS errors related to self-signed certs
            let connector = native_tls::TlsConnector::builder()
                .danger_accept_invalid_certs(true)
                .build()
                .expect("Failed to build TLS connector");

            let connector = tokio_tungstenite::Connector::NativeTls(connector);

            tokio_tungstenite::connect_async_tls_with_config(request, None, false, Some(connector))
                .await
                .expect("Failed to connect (TLS)")
        } else {
            tokio_tungstenite::connect_async(request)
                .await
                .expect("Failed to connect")
        };

        assert_eq!(
            response.headers().get("sec-websocket-protocol"),
            Some(&HeaderValue::from_static(SUBPROTOCOL))
        );

        Self { stream }
    }

    /// Receives a message from the server.
    pub async fn recv_msg(&mut self) -> Result<Message, RecvError> {
        match self.stream.next().await {
            Some(r) => r.map_err(RecvError::from),
            None => Err(RecvError::UnexpectedEndOfStream),
        }
    }

    /// Receives and parses a message from the server.
    pub async fn recv(&mut self) -> Result<ServerMessage, RecvError> {
        let msg = tokio::time::timeout(Duration::from_secs(1), self.recv_msg()).await??;
        let msg = ServerMessage::try_from(&msg)?;
        Ok(msg.into_owned())
    }

    /// Sends a message to the server.
    pub async fn send(&mut self, msg: impl Into<Message>) -> Result<(), tungstenite::Error> {
        self.stream.send(msg.into()).await
    }

    /// Closes the websocket connection.
    pub async fn close(&mut self) {
        self.stream.close(None).await.unwrap();
    }
}
