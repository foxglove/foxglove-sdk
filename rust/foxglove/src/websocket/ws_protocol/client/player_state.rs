use bytes::{Buf, BufMut};

use crate::websocket::ws_protocol::{BinaryMessage, ParseError};

use super::BinaryOpcode;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum PlaybackState {
    Playing = 0,
    Paused = 1,
    Buffering = 2,
    Ended = 3,
}

impl TryFrom<u8> for PlaybackState {
    type Error = u8;

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(Self::Playing),
            1 => Ok(Self::Paused),
            2 => Ok(Self::Buffering),
            3 => Ok(Self::Ended),
            _ => Err(value),
        }
    }
}

/// Player state message.
#[derive(Debug, Clone, PartialEq)]
pub struct PlayerState {
    /// Playback state
    pub playback_state: PlaybackState,
    /// Playback speed
    pub playback_speed: f32,
    /// Seek playback time in nanoseconds (only set if a seek has been performed)
    pub seek_time: Option<u64>,
}

impl<'a> BinaryMessage<'a> for PlayerState {
    fn parse_binary(mut data: &'a [u8]) -> Result<Self, ParseError> {
        // Message size: playback_state (1 byte) + playback_speed (4 bytes) + had_seek (1 byte) + seek_time (8 bytes)
        if data.len() < 1 + 4 + 1 + 8 {
            return Err(ParseError::BufferTooShort);
        }

        let state_byte = data.get_u8();
        let playback_state = PlaybackState::try_from(state_byte)
            .map_err(|_| ParseError::InvalidPlaybackState(state_byte))?;

        let playback_speed = data.get_f32_le();
        let had_seek = data.get_u8() != 0;
        let seek_time = if had_seek && data.len() >= 8 {
            Some(data.get_u64_le())
        } else {
            None
        };

        Ok(Self {
            playback_state,
            playback_speed,
            seek_time,
        })
    }

    fn to_bytes(&self) -> Vec<u8> {
        // Message size: opcode (1) + playback_state (1) + playback_speed (4) + had_seek (1) + seek_time (8)
        let mut buf = Vec::with_capacity(1 + 4 + 1 + 8);

        buf.put_u8(BinaryOpcode::PlayerState as u8);
        buf.put_u8(self.playback_state as u8);
        buf.put_f32_le(self.playback_speed);
        buf.put_u8(if self.seek_time.is_some() { 1 } else { 0 });
        // To keep the message size constant, write the seek time even if it's None.
        buf.put_u64_le(self.seek_time.unwrap_or(0));
        buf
    }
}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::client::ClientMessage;

    use super::*;

    #[test]
    fn test_encode_playing() {
        let message = PlayerState {
            playback_state: PlaybackState::Playing,
            playback_speed: 1.0,
            seek_time: None,
        };
        insta::assert_snapshot!(format!("{:#04x?}", message.to_bytes()));
    }

    #[test]
    fn test_encode_paused() {
        let message = PlayerState {
            playback_state: PlaybackState::Paused,
            playback_speed: 1.0,
            seek_time: None,
        };
        insta::assert_snapshot!(format!("{:#04x?}", message.to_bytes()));
    }

    #[test]
    fn test_encode_buffering() {
        let message = PlayerState {
            playback_state: PlaybackState::Buffering,
            playback_speed: 1.0,
            seek_time: None,
        };
        insta::assert_snapshot!(format!("{:#04x?}", message.to_bytes()));
    }

    #[test]
    fn test_encode_ended() {
        let message = PlayerState {
            playback_state: PlaybackState::Ended,
            playback_speed: 1.0,
            seek_time: None,
        };
        insta::assert_snapshot!(format!("{:#04x?}", message.to_bytes()));
    }

    #[test]
    fn test_roundtrip_playing_with_seek_time() {
        let orig = PlayerState {
            playback_state: PlaybackState::Playing,
            playback_speed: 1.0,
            seek_time: Some(100_500_000_000),
        };
        let buf = orig.to_bytes();
        let msg = ClientMessage::parse_binary(&buf).unwrap();
        assert_eq!(msg, ClientMessage::PlayerState(orig));
    }

    #[test]
    fn test_roundtrip_playing_without_seek_time() {
        let orig = PlayerState {
            playback_state: PlaybackState::Playing,
            playback_speed: 1.0,
            seek_time: None,
        };
        let buf = orig.to_bytes();
        let msg = ClientMessage::parse_binary(&buf).unwrap();
        assert_eq!(msg, ClientMessage::PlayerState(orig));
    }

    #[test]
    fn test_parse_binary_with_seek_time() {
        // Manually construct binary data: opcode + state + speed + had_seek + seek_time
        let mut data = Vec::new();
        data.put_u8(BinaryOpcode::PlayerState as u8); // opcode
        data.put_u8(PlaybackState::Playing as u8); // state
        data.put_f32_le(1.5); // speed
        data.put_u8(1); // had_seek = true
        data.put_u64_le(100_500_000_000); // seek_time

        let msg = ClientMessage::parse_binary(&data).unwrap();
        match msg {
            ClientMessage::PlayerState(state) => {
                assert_eq!(state.playback_state, PlaybackState::Playing);
                assert_eq!(state.playback_speed, 1.5);
                assert_eq!(state.seek_time, Some(100_500_000_000));
            }
            _ => panic!("Expected PlayerState message"),
        }
    }

    #[test]
    fn test_parse_binary_without_seek_time() {
        // Manually construct binary data with had_seek = false (seek_time bytes still present but zeroed)
        let mut data = Vec::new();
        data.put_u8(BinaryOpcode::PlayerState as u8); // opcode
        data.put_u8(PlaybackState::Buffering as u8); // state
        data.put_f32_le(2.0); // speed
        data.put_u8(0); // had_seek = false
        data.put_u64_le(0); // seek_time (zeroed out, ignored since had_seek = false)

        let msg = ClientMessage::parse_binary(&data).unwrap();
        match msg {
            ClientMessage::PlayerState(state) => {
                assert_eq!(state.playback_state, PlaybackState::Buffering);
                assert_eq!(state.playback_speed, 2.0);
                assert_eq!(state.seek_time, None);
            }
            _ => panic!("Expected PlayerState message"),
        }
    }
}
