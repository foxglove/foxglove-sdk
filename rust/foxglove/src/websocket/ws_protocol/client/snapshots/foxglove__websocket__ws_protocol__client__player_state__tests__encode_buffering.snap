---
source: rust/foxglove/src/websocket/ws_protocol/client/player_state.rs
expression: "format!(\"{:#04x?}\", message_buffering().to_bytes())"
---
[
    0x03,
    0x02,
    0x00,
    0x00,
    0x80,
    0x3f,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
]
